Update documentation (#546)
qiboteam/qibolab: doc/source/main-documentation/qibolab.rst

- setter and getter of channel/qubit parameters (local oscillator parameters, attenuations, gain and biases)

The idea of the ``Platform`` is to serve as the only exposed object to the user so, for example, we can easily write an example of experiment, whithout any need of going into the low-level instrument-dpecific code.

For example, let's first define a platform (that we consider to be a single qubit platform) using the ``create`` method presented in :doc:`/tutorials/lab`:

.. code-block::  python

    from qibolab import Platform
    platform = Platform("my_platform")

Now we connect and start the instruments (note that we, the user, do not need to know which instruments are connected).

.. code-block::  python

    platform.connect()
    platform.setup()
    platform.start()

We can easily print some of the parameters of the channels (similarly we can set those, if needed):

.. note::
   If the get_method does not apply to the platform (for example there is no local oscillator, to TWPA or no flux tunability...) a ``NotImplementedError`` will be raised.

.. code-block::  python

>     print(f"Drive LO frequency: {platform.get_lo_drive_frequency(0)}")

Reviewer: My idea was to replace these methods with the qubit/channel API. So instead of
```py
platform.get_lo_drive_frequency(0)
```
use
```py
platform.qubits[0].drive.lo_frequency
```
but I didn't do it yet, as it would require changing qibocal too. 

Since currently both exist it is fine to use either in the documentation. I am in favor of the latter, but that's my personal opinion which we don't necessarily need to follow.

    print(f"Readout LO frequency: {platform.get_lo_readout_frequency(0)}")
    print(f"TWPA LO frequency: {platform.get_lo_twpa_frequency(0)}")
    print(f"TWPA LO power: {platform.get_lo_twpa_power(0)}")

    print(f"Qubit bias: {platform.get_bias(0)}")
    print(f"Qubit attenuation: {platform.get_attenuation(0)}")

> Now we can create a simple sequence (again, without explicitly give any qubit specific parameter, defined in the runcard):

Reviewer: ```suggestion
Now we can create a simple sequence (again, without explicitly giving any qubit specific parameter, as these are loaded automatically from the platform, as defined in the runcard):
```
if I understand it correctly.

.. code-block::  python

   from qibolab.pulses import PulseSequence

   ps = PulseSequence()
   ps.add(platform.create_RX_pulse(qubit=0, start=0))   # start time is in ns
   ps.add(platform.create_RX_pulse(qubit=0, start=100))
   ps.add(platform.create_MZ_pulse(qubit=0, start=200))

Now we can execute the sequence on hardware:

.. code-block::  python

   results = platform.execute_pulse_sequence(ps)

Finally, we can stop instruments and close connections.

.. code-block::  python

    platform.stop()
    platform.disconnect()

Qubits
------

The :class:`qibolab.qubits.Qubit` class serves as a comprehensive representation of a physical qubit within the Qibolab framework.
It encapsulates three fundamental elements crucial to qubit control and operation:

- Channels: Physical Connections
- Parameters: Configurable Properties

> - Native Gates: Quantum Operations

Reviewer: Also here we could link to the relevant sections.

Channels play a pivotal role in connecting the quantum system to the control infrastructure.
They are optional and encompass distinct types, each serving a specific purpose:

- readout (from controller device to the qubits)
- feedback (from qubits to controller)

> - twpa (pump to the twpa)

Reviewer: ```suggestion
- twpa (pump to the TWPA)
```

- drive
- flux
- flux_coupler

The Qubit class allows you to set and manage several key parameters that influence qubit behavior.
These parameters are typically extracted from the runcard during platform initialization.

> Notable settable parameters include:

Reviewer: I am not sure if we really need to list all these here, since they should be (if we have the proper docstrings) in the API reference.

The controller's driver ensures the correct pulse frequency is set based on the LO's configuration.

Let's explore an example using an RFSoC controller.
Note that while channels are defined in a device-independent manner, the port parameter varies based on the specific instrument.

.. code-block:: python

    from qibolab.channels import Channel, ChannelMap
    from qibolab.instruments.rfsoc import RFSoC

    controller = RFSoC()
    channel1 = Channel("my_channel_name_1", port=controller[1])
    channel2 = Channel("my_channel_name_2", port=controller[2])
    channel3 = Channel("my_channel_name_3", port=controller[3])

Channels are then organized in :class:`qibolab.channels.ChannelMap` to be passed as a single argument to the platform.
Following the tutorial in :doc:`/tutorials/lab`, we can continue the initialization:

.. code-block:: python

    ch_map = ChannelMap()
    ch_map |= channel1
    ch_map |= channel2
    ch_map |= channel3

    platform = Platform(Name, runcard, instruments, ch_map)

    platform.qubits[0].drive = channel1
    platform.qubits[0].readout = channel2

>     platform.qubits[0].feedback = channel3

Reviewer: Also this will be different after #525.

Qibolab offers a range of pre-defined pulse shapes:

- Rectangular (:class:`qibolab.pulses.Rectangular`)
- Exponential (:class:`qibolab.pulses.Exponential`)
- Gaussian (:class:`qibolab.pulses.Gaussian`)
- Drag (:class:`qibolab.pulses.Drag`)
- IIR (:class:`qibolab.pulses.IIR`)
- SNZ (:class:`qibolab.pulses.SNZ`)
- eCap (:class:`qibolab.pulses.eCap`)
- Custom (:class:`qibolab.pulses.Custom`)

To illustrate, here are some examples of single pulses using the Qibolab API:

.. code-block:: python

    from qibolab.pulses import Pulse, Rectangular

    pulse = Pulse(
        start=0,  # Timing, always in nanoseconds (ns)
        duration=40,  # Pulse duration in ns
        amplitude=0.5,  # Amplitude relative to instrument range
        frequency=1e8,  # Frequency in Hz
        relative_phase=0,  # Phase in radians
        shape=Rectangular(),
        channel="channel",
        type="qd",  # Enum type: :class:`qibolab.pulses.PulseType`
        qubit=0,
    )

In this way, we defined a rectangular drive pulse using the generic Pulse object.
Alternatively, you can achieve the same result using the dedicated :class:`qibolab.pulses.DrivePulse` object:

.. code-block:: python

>     from qibolab.pulses import *

Reviewer: ```suggestion
    from qibolab.pulses import DrivePulse, Rectangular
```

Lastly, when conducting an experiment, it is not always required to define a pulse from scratch.
Usual pulses, as pi-pulses or measurements are already defined in the platform runcard and can be easily initialized with platform methods.
Typical experiments may include both pre-defined pulses and new ones:

.. code-block:: python

    sequence = PulseSequence()
    sequence.add(my_platform.create_RX_pulse())
    sequence.add(DrivePulse(...))
    sequence.add(my_platform.create_MZ_pulse())

    results = my_platform.execute_pulse_sequence(sequence, options=options)

.. note::

   options is an :class:`qibolab.execution_parameters.ExecutionParameters` object, detailed in a separate section.

Symbolic expressions
--------------------

Sweepers
--------

Sweeper objects, represented by the :class:`qibolab.sweeper.Sweeper` class, stand as a crucial component in experiments and calibration tasks within the Qibolab framework.

Consider a scenario where a resonator spectroscopy experiment is performed. This process involves a sequence of steps:

1. Define a pulse sequence.
2. Define a readout pulse with frequency A.
3. Execute the sequence.

> 4. Define a new readout pulse with frequency A + :math:`\epsilon`.

Reviewer: Minor: these equations look a bit weird because A is just a letter and \epsilon looks like it's compiled to an image. Maybe it would be better to use one of the two approaches.

A tipical resonator spectroscopy experiment could be defined with:

.. code-block:: python

    sequence = PulseSequence()
    sequence.add(my_platform.create_MZ_pulse(0))  # readout pulse for qubit 0 at 4 GHz
    sequence.add(my_platform.create_MZ_pulse(1))  # readout pulse for qubit 1 at 5 GHz
    sequence.add(my_platform.create_MZ_pulse(2))  # readout pulse for qubit 2 at 6 GHz

    sweeper = Sweeper(
        parameter=Parameter.frequency,
        values=np.arange(-200_000, +200_000, 1),  # define an interval of swept values
        pulses=[sequence[0], sequence[1], sequence[2]],
        type=SweeperType.OFFSET,
    )

    results = my_platform.sweep(sequence, options, sweeper)

.. note::

   options is an :class:`qibolab.execution_parameters.ExecutionParameters` object, detailed in a separate section.

In this way, we first define a sweeper with an interval of 400 MHz (-200 MHz --- 200 MHz), assigning it to all three readout pulses and setting is as an offset sweeper. The resulting probed frequency will then be:
    - for qubit 0: [3.8 GHz, 4.2 GHz]
    - for qubit 1: [4.8 GHz, 5.2 GHz]
    - for qubit 2: [5.8 GHz, 6.2 GHz]

> If we had used the SweeperType absolute, we would have probed for all qubits the same frequencies [-200 MHz, 200 MHz].

Reviewer: ```suggestion
If we had used the :class:`qibolab.sweeper.SweeperType` absolute, we would have probed for all qubits the same frequencies [-200 MHz, 200 MHz].
```

.. note::

>    The default SweeperType is absolute!

Reviewer: ```suggestion
   The default :class:`qibolab.sweeper.SweeperType` is absolute!
```

Let's now delve into the details of the ``options`` parameter and understand its components.

The ``options`` parameter, represented by the :class:`qibolab.execution_parameters.ExecutionParameters` class, is a vital element for every hardware execution. It encompasses essential information that tailors the execution to specific requirements:

- ``nshots``: Specifies the number of experiment repetitions.
- ``relaxation_time``: Introduces a wait time between repetitions, measured in nanoseconds (ns).
- ``fast_reset``: Enables or disables fast reset functionality, if supported; raises an error if not supported.
- ``acquisition_type``: Determines the acquisition mode for results.
- ``averaging_mode``: Defines the mode for result averaging.

The first three parameters are straightforward in their purpose. However, let's take a closer look at the last two parameters.

Supported acquisition types, accessible via the :class:`qibolab.execution_parameters.AcquisitionType` enumeration, include:

- Discrimination: Distinguishes states based on acquired voltages.
- Integration: Returns demodulated and integrated waveforms.
- Raw: Offers demodulated, yet unintegrated waveforms.

> - Spectroscopy: Designed specifically for spectroscopy applications, primarily supported by Zurich Instruments devices.

Reviewer: Shall we mention "Spectroscopy" here? I would skip since it is not relevant for any instrument other than Zurich and I am not even sure if the code supports it.

Zurich Instruments
^^^^^^^^^^^^^^^^^^

> ``Qibolab`` has been tested with the following `instrument cluster <https://www.zhinst.com/others/en/instruments/product-finder/type/quantum_computing_systems>`_:

Reviewer: I was generally using Qibolab instead of ``Qibolab`` across the documentation, including the QM section right below this, but I am not sure if we have decided on a convention.
